### PR TITLE
[Doc] EZP-27294: Add option to declare list of authors from which the tweets are allowed

### PR DESCRIPTION
**Pull request by jacek-foremski**
Documentation changes reflecting code changes from #28.

>JIRA: [EZP-27294](https://jira.ez.no/browse/EZP-27294)
This PR aims at adding the possibility to declare a list of allowed tweet authors.
This feature currently is signalled throughout the code and in the tutorial documentation but is not implemented fully. My PR adds the missing pieces.

### CODE REVIEW

File: docs/add_a_validation.md [hunk 1]
```diff
@@ -0,0 +1,295 @@
+# Add a validation
+
+We want to add the option to configure list of authors from which the tweets are allowed. To achieve this, we have to:
```
Comment (DominikaK) [inline docs/add_a_validation.md]: We want to add the option to configure **a** list of authors **whose** tweets are allowed.

File: docs/add_a_validation.md [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+
+private function isAuthorApproved($author, $validatorConfiguration)
+{
+    return !isset($validatorConfiguration['TweetValueValidator'])
+        || empty($validatorConfiguration['TweetValueValidator']['authorList'])
+        || in_array($author, $validatorConfiguration['TweetValueValidator']['authorList']);
+}
+```
+
+In addition to validating the URL with a regular expression, which we implemented before, if your Field Type instance’s configuration contains a `TweetValueValidator` key, you will check that the username in the status URL matches one of the valid authors.
```
Comment (DominikaK) [inline docs/add_a_validation.md]: A bit longing, I'd rephrase it like this:
Earlier we validated the URL with a regular expression. Now, if your the configuration of your Field Type's instance contains a `TweetValueValidator` key, you will check that the username in the status URL matches one of the valid authors.

File: docs/add_a_validation.md [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+
+## Implement FormMapper
+
+Next, we need to create a way for user to input a list of authors for our validation. To achieve this, we need to implement a FormMapper that allows us to add an input field to form with content field definition.
```
Comment (DominikaK) [inline docs/add_a_validation.md]: Next, we need to create a way for **the** user to input a list of authors for our validation. To achieve this, we need to implement a FormMapper that allows us to add an input field to **the form where the user defines the fields**.

File: docs/add_a_validation.md [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+    }
+}
+```
+
+Unfortunately, because in our case authorList validator has array type, we cannot leave it like that. We will need a way to transform data from array to comma-separated string and vice versa.
+To achieve this, first we need to implement a DataTransformer:
+
+```php
+// Form/StringToArrayTransformer.php
+
```
Comment (DominikaK) [inline docs/add_a_validation.md]: Wouldn't it be better to put the full file here? An over-eager copy-paster (like me) could overlook the first 15 lines and get errors (unless you consider it to be an obvious step for someone who uses this tutorial)

File: docs/add_a_validation.md [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+        return array_map('trim', $array);
+    }
+}
+```
+
+Then, we can use this DataTransformer in our FormMapper like this:
+```php
+// eZ/Publish/FieldType/Tweet/FormMapper.php
+
```
Comment (DominikaK) [inline docs/add_a_validation.md:220]: same comment as above

File: docs/add_a_validation.md [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+    ezsystems.tweetbundle.fieldtype.eztweet.form_mapper:
+        class: EzSystems\TweetFieldTypeBundle\eZ\Publish\FieldType\Tweet\FormMapper
+        tags:
+            - {name: ez.fieldFormMapper.definition, fieldType: eztweet}
+```
+
+## Add field definition edit view
+
+We have the new part of the form defined, but we still need to show it to the user. To do that, we need to create file containing the view:
```
Comment (DominikaK) [inline docs/add_a_validation.md]: To do that, we need to create **a** file containing the view:

File: docs/add_a_validation.md [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+                'authorList' => $authorList
+            ],
+        ];
+    }
+}
+```
+
+`toStorageFieldDefinition()` converts a field definition to a legacy one using the `dataText1` field (which in this example means converting it to json), and `toFieldDefinition()` converts a stored legacy field definition to an API field definition (which means converting it back to an array according to validation schema).
+
+You should now be able to configure list of authors when editing Content Type with Tweet Field Type.
```
Comment (DominikaK) [inline docs/add_a_validation.md]: You should now be able to configure **a** list of authors when editing **a** Content Type with **the** Tweet Field Type.

**Review by DominikaK — CHANGES_REQUESTED** [review]
A few small clarifications in the comments

**Comment by jacek-foremski** [comment]
I have added your fixes. Thanks a lot @DominikaK :)

**Review by DominikaK — APPROVED** [review]

File: docs/add_a_validation.md [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+    'TweetValueValidator' => [
+        'authorList' => [
+            'type' => 'array',
+            'default' => []
+        ]
+    ]
+];
+```
+
+For a TextLine (length validation), it means checking that both min length and max length are positive integers, and that min is lower than max.
```
Comment (mere-teresa) [inline docs/add_a_validation.md]: Maybe put a First Uppercase to Min Length et Max Length to reinforce the information that they are Fields? This sentence is very hard to read.

File: docs/add_a_validation.md [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+
+When an instance of the type is added to a Content Type, `validateValidatorConfiguration()` receives the configuration for the validators used by the Type as an array. It must return an array of error messages if errors are found in the configuration, and an empty array if no errors were found.
```
Comment (mere-teresa) [inline docs/add_a_validation.md]: > It must return 

Replace it by the real subject of the sentence, please. We would like to ease the reading.

File: docs/add_a_validation.md [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+
+For TextLine, the provided array looks like this:
```
Comment (mere-teresa) [inline docs/add_a_validation.md]: I propose:

> For the Fieldtype TextLine, the array provided by ... looks like this:

File: docs/add_a_validation.md [hunk 11]
```diff
@@ -0,0 +1,317 @@
...
+
+array(
+   'StringLengthValidator' => array(
+       'minStringLength' => 0,
+       'maxStringLength' => 100
+   )
+);
+```
+
+The structure of this array is totally free, and up to each type implementation. In this tutorial it will mimic what is done in native Field Types:
```
Comment (mere-teresa) [inline docs/add_a_validation.md]: > The structure of this array is totally free, and up to each type implementation

Is it free or up to each type?
If it's totally free, we could still have best practices or recommendations.

File: docs/add_a_validation.md [hunk 12]
```diff
@@ -0,0 +1,317 @@
...
+
+Each level one key is the name of a validator, as acknowledged by the Type. That key contains a set of parameter name / parameter value rows. You must check that:
+
+- all the validators in this array are known to the type
+- arguments for those validators are valid and have sane values
```
Comment (mere-teresa) [inline docs/add_a_validation.md]: same values than what?

File: docs/add_a_validation.md [hunk 13]
```diff
@@ -0,0 +1,317 @@
...
+        'authorList' => ['johndoe', 'janedoe']
+    ]
+];
+```
+
+The configuration says that tweets must be either by `johndoe` or by `janedoe`. If you had not provided `TweetValueValidator` at all, it would have been ignored.
+
+You will iterate over the items in `$validatorConfiguration` and:
+
+- add errors for those you don’t know about;
```
Comment (mere-teresa) [inline docs/add_a_validation.md]: (for unexpected authors?)

File: docs/add_a_validation.md [hunk 14]
```diff
@@ -0,0 +1,317 @@
...
+        || empty($validatorConfiguration['TweetValueValidator']['authorList'])
+        || in_array($author, $validatorConfiguration['TweetValueValidator']['authorList']);
+}
+```
+
+Earlier we validated the URL with a regular expression. Now, if the configuration of your Field Type's instance contains a TweetValueValidator key, you will check that the username in the status URL matches one of the valid authors.
+
+## Implement FormMapper
+
+Next, we need to create a way for the user to input a list of authors for our validation. To achieve this, we need to implement a FormMapper that allows us to add an input field to the form where the user defines the fields.
```
Comment (mere-teresa) [inline docs/add_a_validation.md]: These sentences are full of doubt to me.
_We would like to offer a way to the user to input a list of authors that will be upon validation. To achieve this, we will implement a FormMapper that allows us to define the input field into the form where the user has defined the fields._

(If I am not correct in English nor in the programming logic, please adapt and correct).

File: docs/add_a_validation.md [hunk 15]
```diff
@@ -0,0 +1,317 @@
...
+The most basic one looks like the following:
```
Comment (mere-teresa) [inline docs/add_a_validation.md]: I propose to change it by:

_This is a minimal example of our FormMapper:_

File: docs/add_a_validation.md [hunk 16]
```diff
@@ -0,0 +1,317 @@
...
+            ->add('authorList', TextType::class, [
+                'required' => false,
+                'property_path' => 'validatorConfiguration[TweetValueValidator][authorList]',
+                'label' => 'field_definition.eztweet.authorList'
+            ]);
+    }
+}
+```
+
+Unfortunately, because in our case authorList validator has array type, we cannot leave it like that. We will need a way to transform data from array to comma-separated string and vice versa.
```
Comment (mere-teresa) [inline docs/add_a_validation.md]: It's not fate-programming, we chose what we are doing. I would like to rephrase it like this:
_In our case, authorList is an array and the validator should be set accordingly. To transform data from array to a comma-separated string and in the other way, we need to implement a DataTransformer:_

File: docs/add_a_validation.md [hunk 17]
```diff
@@ -0,0 +1,317 @@
...
+                    ])
+                    ->addModelTransformer(new StringToArrayTransformer())
+                    // Deactivate auto-initialize as we're not on the root form.
+                    ->setAutoInitialize(false)->getForm()
+            );
+    }
+}
+```
+
+Next thing is to register the FormMapper as a service, so the system would know to use it. To achieve this, let's add the following lines to `fieldtypes.yml`:
```
Comment (mere-teresa) [inline docs/add_a_validation.md]: Add a link to the Services documentation page, here, please.
https://doc.ezplatform.com/en/1.11/guide/service_container/

File: docs/add_a_validation.md [hunk 18]
```diff
@@ -0,0 +1,317 @@
...
+{% block eztweet_field_definition_edit %}
+    <div class="eztweet-validator author_list{% if group_class is not empty %} {{ group_class }}{% endif %}">
+        {{- form_label(form.authorList) -}}
+        {{- form_errors(form.authorList) -}}
+        {{- form_widget(form.authorList) -}}
+    </div>
+{% endblock %}
+```
+
+Also, we need to register it by editing the file `ez_field_templates.yml`:
```
Comment (mere-teresa) [inline docs/add_a_validation.md]: Register it as what?
_Register the new template in the configuration_

File: docs/add_a_validation.md [hunk 19]
```diff
@@ -0,0 +1,317 @@
...
+        ];
+    }
+}
+```
+
+`toStorageFieldDefinition()` converts a field definition to a legacy one using the `dataText1` field (which in this example means converting it to json), and `toFieldDefinition()` converts a stored legacy field definition to an API field definition (which means converting it back to an array according to validation schema).
+
+You should now be able to configure a list of authors when editing a Content Type with the Tweet Field Type.
+You should also have your tweets validated in accordance to this list when you create or edit Content Item with this Field Type. 
+
```
Comment (mere-teresa) [inline docs/add_a_validation.md:313]: In general, I prefer to read "we will" than "we need".

**Review by mere-teresa — CHANGES_REQUESTED** [review]

**Comment by jacek-foremski** [comment]
@mere-teresa I made the changes that you requested. Thanks a lot for your input!

**Review by mere-teresa — APPROVED** [review]
